tokio-boring/src/async_callbacks.rs

```diff
@@ -0,0 +1,262 @@
+use boring::ex_data::Index;
+use boring::ssl::{self, ClientHello, PrivateKeyMethod, Ssl, SslContextBuilder};
+use once_cell::sync::Lazy;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{ready, Context, Poll, Waker};
+
+type BoxSelectCertFuture = ExDataFuture<Result<BoxSelectCertFinish, AsyncSelectCertError>>;
+
+type BoxSelectCertFinish = Box<dyn FnOnce(ClientHello<'_>) -> Result<(), AsyncSelectCertError>>;
+
+/// The type of futures returned by [`AsyncPrivateKeyMethod`] methods.
+pub type BoxPrivateKeyMethodFuture =
+    ExDataFuture<Result<BoxPrivateKeyMethodFinish, AsyncPrivateKeyMethodError>>;
+
+/// The type of callbacks returned by [`BoxPrivateKeyMethodFuture`].
+pub type BoxPrivateKeyMethodFinish =
+    Box<dyn FnOnce(&mut ssl::SslRef, &mut [u8]) -> Result<usize, AsyncPrivateKeyMethodError>>;
+
+type ExDataFuture<T> = Pin<Box<dyn Future<Output = T> + Send + Sync>>;
+
+pub(crate) static TASK_WAKER_INDEX: Lazy<Index<Ssl, Option<Waker>>> =
+    Lazy::new(|| Ssl::new_ex_index().unwrap());
+pub(crate) static SELECT_CERT_FUTURE_INDEX: Lazy<Index<Ssl, BoxSelectCertFuture>> =
+    Lazy::new(|| Ssl::new_ex_index().unwrap());
+pub(crate) static SELECT_PRIVATE_KEY_METHOD_FUTURE_INDEX: Lazy<
+    Index<Ssl, BoxPrivateKeyMethodFuture>,
+> = Lazy::new(|| Ssl::new_ex_index().unwrap());
+
+/// Extensions to [`SslContextBuilder`].
+///
+/// This trait provides additional methods to use async callbacks with boring.
+pub trait SslContextBuilderExt: private::Sealed {
+    /// Sets a callback that is called before most [`ClientHello`] processing
+    /// and before the decision whether to resume a session is made. The
+    /// callback may inspect the [`ClientHello`] and configure the connection.
+    ///
+    /// This method uses a function that returns a future whose output is
+    /// itself a closure that will be passed [`ClientHello`] to configure
+    /// the connection based on the computations done in the future.
+    ///
+    /// See [`SslContextBuilder::set_select_certificate_callback`] for the sync
+    /// setter of this callback.
+    fn set_async_select_certificate_callback<Init, Fut, Finish>(&mut self, callback: Init)
+    where
+        Init: Fn(&mut ClientHello<'_>) -> Result<Fut, AsyncSelectCertError> + Send + Sync + 'static,
+        Fut: Future<Output = Result<Finish, AsyncSelectCertError>> + Send + Sync + 'static,
+        Finish: FnOnce(ClientHello<'_>) -> Result<(), AsyncSelectCertError> + 'static;
```

Comment:
Maybe I'm missing something, but I think the plan was to pass `&mut SslRef` instead of `ClientHello` to the closure? So we can actually change the various options on `SslRef`.

Also, not sure I understand why `Init` takes a `&mut ClientHello` rather than just `&ClientHello`? That doesn't need to be mutable.

Reply:
`ClientHello<'_>` and not `&mut SslRef` because there are methods that are only available on `ClientHello<'_>` (see the existing sync of the callback).

`&mut ClientHello<'_>` and not `&ClientHello<'_>` because getting a `&mut SslRef` from a `&ClientHello<'_>` is very much not good.

Reply:
Just to be extra clear: `ClientHello<'_>` is equipped with `fn ssl_mut(&mut self) -> &mut SslRef` in this PR.

Reply:
Oh, indeed I missed the `ssl_mut()` change, thanks. I understand now.